### ISSUE TITLE
systemd-logind crashes when closing lid after "systemctl mask sleep.target"

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#7466]
### Submission type

 - Bug report

### systemd version the issue has been seen with

v235
git master (v235-542-g2482f88d2)

### Used distribution

Debian
Downstream bug report https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=882270

I want to be able to configure if my laptop should suspend when closing the
lid. Because I do not want to kill the X session for this, changing logind.conf
and restarting logind is not an option. Therefore, I set
HandleLidSwitch=suspend and HandleLidSwitchDocked=suspend and want to manually
disable suspend using "systemctl mask sleep.target suspend.target".

It works insofar that when closing the lid, the laptop doesn't suspend,
but logind crashes multiple times in rapid succession. journalctl:

```
Nov 20 21:39:26 litterbox systemd-logind[10697]: Lid closed.
Nov 20 21:39:26 litterbox systemd-logind[10697]: Suspending...
Nov 20 21:39:26 litterbox systemd-logind[10697]: Assertion 'signal_name[w]' failed at ../src/login/logind-dbus.c:1495, function send_prepare_for(). Aborting.
[...unrelated kernel log entries...]
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Main process exited, code=killed, status=6/ABRT
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Failed with result 'signal'.
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Service has no hold-off time, scheduling restart.
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Scheduled restart job, restart counter is at 2.
Nov 20 21:39:26 litterbox systemd[1]: Stopped Login Service.
Nov 20 21:39:26 litterbox systemd[1]: Starting Login Service...
Nov 20 21:39:26 litterbox systemd[1]: Started Login Service.
Nov 20 21:39:26 litterbox systemd-logind[10892]: New seat seat0.
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event6 (Power Button)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event4 (Power Button)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event3 (Lid Switch)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event5 (Sleep Button)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event1 (Logitech USB Receiver)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event2 (Logitech USB Receiver)
Nov 20 21:39:26 litterbox systemd-logind[10892]: Watching system buttons on /dev/input/event0 (AT Translated Set 2 keyboard)
Nov 20 21:39:26 litterbox systemd-logind[10892]: New session 4 of user penma.
Nov 20 21:39:26 litterbox systemd-logind[10892]: New session 1 of user penma.
Nov 20 21:39:26 litterbox systemd-logind[10892]: Suspending...
Nov 20 21:39:26 litterbox systemd-logind[10892]: Assertion 'signal_name[w]' failed at ../src/login/logind-dbus.c:1495, function send_prepare_for(). Aborting.
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Main process exited, code=killed, status=6/ABRT
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Failed with result 'signal'.
Nov 20 21:39:26 litterbox systemd[1]: systemd-logind.service: Service has no hold-off time, scheduling restart.
[...continues multiple times until...]
Nov 20 21:39:27 litterbox systemd[1]: systemd-logind.service: Scheduled restart job, restart counter is at 7.
Nov 20 21:39:27 litterbox systemd[1]: Stopped Login Service.
Nov 20 21:39:27 litterbox systemd[1]: systemd-logind.service: Start request repeated too quickly.
Nov 20 21:39:27 litterbox systemd[1]: systemd-logind.service: Failed with result 'signal'.
Nov 20 21:39:27 litterbox systemd[1]: Failed to start Login Service.
```
after which the whole X session goes down because
```
[  1689.396] (EE)
Fatal server error:
[  1689.396] (EE) systemd-logind disappeared (stopped/restarted?)
[  1689.396] (EE)
[  1689.396] (EE)
Please consult the The X.Org Foundation support at http://wiki.x.org for help.
```
The crashes are reproducible when the events above are masked, and they
also reliably disappear after unmasking said events.

I feel that logind should not crash in this situation, even if there happens to
be a better way to prevent suspend from happening. (For what I'm trying to
achieve right now, "systemd-inhibit --what=handle-lid-switch cat" is good
enough and doesn't crash logind)

### In case of bug report: Steps to reproduce the problem
```systemctl mask sleep.target suspend.target```
Then close lid and open it again

**Comment by mbiebl** [comment]
I can reproduce the issue as well. journal and status output are attached.
[journal.txt](https://github.com/systemd/systemd/files/1503000/journal.txt)
[status.txt](https://github.com/systemd/systemd/files/1503002/status.txt)

**Comment by poettering** [comment]
please provide a backtrace for all crashes like this! systemd-coredump is your friend

**Comment by mbiebl** [comment]
[trace.txt](https://github.com/systemd/systemd/files/1503106/trace.txt)
trace gathered via ```coredumpctl info /lib/systemd/systemd-logind```

**Comment by mbiebl** [comment]
[gdb.txt](https://github.com/systemd/systemd/files/1503330/gdb.txt)

**Comment by poettering** [comment]
Fix waiting in #7475. PTAL and test!
